### PR TITLE
refactor(resolve): splitting glob and non-glob name bindings

### PR DESCRIPTION
**Pull request by bvanjoi**
Fixes this [comment](https://github.com/bvanjoi/rust/blob/f7330eb752cdd83a9b944d1d64a092c7d2454c17/compiler/rustc_resolve/src/ident.rs#L873-L876).

r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
Meta-comment: could you avoid changing and moving code (to `compiler/rustc_resolve/src/name_resolution.rs`) in a same commit?
Let's start with the changes only, and then the move can be done later in a separate commit (or not done at all).

**Comment by bvanjoi** [comment]
> the move can be done later in a separate commit (or not done at all).

The reason I used a segregation file was to hide the `﻿non_glob_binding` and `﻿glob_binding` fields. However, I will move them back for a better review process.

**Comment by bvanjoi** [comment]
All tests had passed, @rustbot ready

**Comment by petrochenkov** [comment]
scopes

**Comment by petrochenkov** [comment]
This PR looks like it is in the right direction, but it is not exactly what I wanted to do.

The picture above shows how non-glob and glob names in a module should be integrated into name lookup as two different scopes.
In that case errors like `GlobVsExpanded` would be subsumed by more general errors like `MoreExpandedVsOuter` (because globs would be an outer scope), `ignore_binding` would be treated more consistently, and any other ad hoc treatment of non-glob and glob names would be replaced with regular scope walk.

I'm not sure whether non-glob and glob resolutions for the same name should be packed into the same `NameResolution`, maybe we should have separate `module.non_glob_resolutions` and `module.glob_resolutions` instead, that would depend on convenience of implementation.

File: compiler/rustc_resolve/src/imports.rs [hunk 1]
```diff
@@ -212,25 +210,32 @@ pub(crate) struct NameResolution<'a> {
...
-    /// Returns the binding for the name if it is known or None if it not known.
     pub(crate) fn binding(&self) -> Option<&'a NameBinding<'a>> {
-        self.binding.and_then(|binding| {
-            if !binding.is_glob_import() || self.single_imports.is_empty() {
-                Some(binding)
-            } else {
-                None
-            }
+        self.non_glob_binding()
+            .or_else(|| if self.single_imports.is_empty() { self.glob_binding() } else { None })
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/imports.rs:220]: This logic is likely incorrect and is a source of issues like https://github.com/rust-lang/rust/issues/56593 (because it doesn't account for macros), but let's keep it as is in this PR.

I'd just want to rename it to something more explanatory, like `determinate_binding`, keep the removed comment and add a FIXME.

File: compiler/rustc_resolve/src/imports.rs [hunk 2]
```diff
@@ -212,25 +210,32 @@ pub(crate) struct NameResolution<'a> {
...
+    pub(crate) fn glob_binding(&self) -> Option<&'a NameBinding<'a>> {
+        self.glob_binding.and_then(|binding| {
+            assert!(binding.is_glob_import());
+            Some(binding)
         })
     }
 
-    pub(crate) fn add_single_import(&mut self, import: &'a Import<'a>) {
-        self.single_imports.insert(Interned::new_unchecked(import));
+    pub(crate) fn available_binding(&self) -> Option<&'a NameBinding<'a>> {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/imports.rs:237]:
```suggestion
    pub(crate) fn binding(&self) -> Option<&'a NameBinding<'a>> {
```
Assuming the other method is renamed to `determinate_binding`.

File: compiler/rustc_resolve/src/imports.rs [hunk 3]
```diff
@@ -212,25 +210,32 @@ pub(crate) struct NameResolution<'a> {
...
-            } else {
-                None
-            }
+        self.non_glob_binding()
+            .or_else(|| if self.single_imports.is_empty() { self.glob_binding() } else { None })
+    }
+
+    pub(crate) fn non_glob_binding(&self) -> Option<&'a NameBinding<'a>> {
+        self.non_glob_binding.and_then(|binding| {
+            assert!(!binding.is_glob_import());
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/imports.rs:225]: I'm not sure these asserts are useful, there's literally one place where these fields are assigned and the glob-ness is checked there.
(I guess they could be kept during development and removed before merging.)

File: compiler/rustc_resolve/src/imports.rs [hunk 4]
```diff
@@ -769,9 +750,8 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                                 .emit();
                         }
                         let key = BindingKey::new(target, ns);
-                        this.update_resolution(parent, key, |_, resolution| {
-                            resolution.single_imports.remove(&Interned::new_unchecked(import));
-                        });
+                        let mut resolution = this.resolution(parent, key).borrow_mut();
+                        resolution.single_imports.remove(&Interned::new_unchecked(import));
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/imports.rs:754]: `update_resolution` has some extra logic related to `glob_importers`.
Does it never run in this case? Or it runs but it's irrelevant?

File: compiler/rustc_resolve/src/imports.rs [hunk 5]
```diff
@@ -305,65 +310,41 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
...
-                                resolution.shadowed_glob = Some(glob_binding);
-                            }
-                        } else {
-                            resolution.shadowed_glob = Some(glob_binding);
-                        }
-                    }
-                    (false, false) => {
-                        return Err(old_binding);
-                    }
+        if let Some(old_binding) = resolution.available_binding() && res == Res::Err && old_binding.res() != Res::Err {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/imports.rs:313]: Could you keep the old code structure here when possible?
Otherwise it's hard to see what changes here.

File: compiler/rustc_resolve/src/build_reduced_graph.rs [hunk 6]
```diff
@@ -378,7 +379,7 @@ impl<'a, 'b, 'tcx> BuildReducedGraphVisitor<'a, 'b, 'tcx> {
                     if !type_ns_only || ns == TypeNS {
                         let key = BindingKey::new(target, ns);
                         let mut resolution = this.resolution(current_module, key).borrow_mut();
-                        resolution.add_single_import(import);
+                        resolution.single_imports.insert(Interned::new_unchecked(import));
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/build_reduced_graph.rs:382]: Nit: unrelated change, could be moved to a separate cleanup commit.

File: compiler/rustc_resolve/src/ident.rs [hunk 7]
```diff
@@ -870,16 +870,11 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         let resolution =
             self.resolution(module, key).try_borrow_mut().map_err(|_| (Determined, Weak::No))?; // This happens when there is a cycle of imports.
 
-        // If the primary binding is unusable, search further and return the shadowed glob
-        // binding if it exists. What we really want here is having two separate scopes in
-        // a module - one for non-globs and one for globs, but until that's done use this
-        // hack to avoid inconsistent resolution ICEs during import validation.
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/ident.rs:876]: Could you keep this comment?
This PR doesn't introduce two separate scopes yet.

File: compiler/rustc_resolve/src/ident.rs [hunk 8]
```diff
@@ -900,7 +895,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             }
 
             // Forbid expanded shadowing to avoid time travel.
-            if let Some(shadowed_glob) = resolution.shadowed_glob
+            if let Some(shadowed_glob) = resolution.glob_binding()
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/ident.rs:898]: This line alone looks like a change in semantics - the binding is still glob now, but not necessarily shadowed.
But the `binding.res() != shadowed_glob.res()` condition below should ensure the old behavior, right?

File: tests/ui/resolve/non-glob-first.rs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    }
+}
+pub use b::P;
+
+mod c {
+    use crate::*;
+    pub struct _S(Vec<P>);
+}
+
+fn main() {}
```
Comment (petrochenkov) [inline tests/ui/resolve/non-glob-first.rs:21]: Is there some specific issue for this test?

**Comment by bvanjoi** [comment]
> it is not exactly what I wanted to do.

Apologize for the misunderstanding on this [comment](https://github.com/bvanjoi/rust/blob/f7330eb752cdd83a9b944d1d64a092c7d2454c17/compiler/rustc_resolve/src/ident.rs#L873-L876). I just think `shaow_glob` field to be a hack that should be removed, so I was splitting it into two separate fields.

> In that case errors like ....

It appears that simply splitting ﻿`non-glob-binding` and ﻿`glob-binding` fields will not directly solve the problem.

> we should have separate module.non_glob_resolutions and module.glob_resolutions

It appears that we may be able to delete the `﻿shadow_glob` field in `NameResolution` once we have separated ﻿`resolver.non_glob_resolution(module)` and `﻿resolver.glob_resolutions(module)`. I will investigate this further soon.

File: compiler/rustc_resolve/src/imports.rs (hunk 1 above)
Comment (bvanjoi) [reply compiler/rustc_resolve/src/imports.rs:220]: Yes, I think that issue #56593 is interesting and could be the next task to work on.

File: compiler/rustc_resolve/src/imports.rs (hunk 4 above)
Comment (bvanjoi) [reply compiler/rustc_resolve/src/imports.rs:754]: It is a cleanup, because it seems will never meet other logic in `update_resolution` because the `binding` is not updated, and it will hit following code always:

```rust
 match resolution.binding() {
           _ if old_binding.is_some() => return t,
          None => return t,
      // xxxxxx
}
```


Of course, It should revert it back because it is irrrelative.

File: tests/ui/resolve/non-glob-first.rs (hunk 9 above)
Comment (bvanjoi) [reply tests/ui/resolve/non-glob-first.rs:21]: No specific problem, it just `tests/ui` passed after refactoring but block by this [comment](https://github.com/rust-lang/rust/pull/112659#issuecomment-1593369350). So I reduced it and added it into `tests/ui` after fixed.

btw, It is had some relative with #56593

**Comment by bvanjoi** [comment]
I should clarify that I will not fix these comments temporarily because this refactoring does not appear to address the underlying issue.

**Comment by JohnCSimon** [comment]
@bvanjoi 

ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by bvanjoi** [comment]
No updates for now, I'll look at this after I fix the other issues, so close it for now.
